server/src/main/java/org/opensearch/common/concurrent/GatedCloseable.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.common.concurrent;
+
+import org.opensearch.common.CheckedRunnable;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/**
+ * Decorator class that wraps an object reference with a {@link CheckedRunnable} that is
+ * invoked when {@link #close()} is called. The internal {@link OneWayGate} instance ensures
+ * that this is invoked only once. See also {@link GatedAutoCloseable}
+ */
+public class GatedCloseable<T> implements Closeable {
```

Comment:
Wish there was a way to combine this further with the AutoCloseable class.  Then we wouldn't need the OneWayGate class.  I think the limitation is CheckedRunnable does not inherit from Runnable?

Reply:
It's also worth noting that the method signature for `run()` in `Closeable` throws a checked exception (`IOException`) while the one in `AutoCloseable` simply throws an exception. A unified `Gated*` class would necessarily need to throw checked exceptions, introducing exception handling in places that didn't need to handle it before.

So to avoid those changes, i've opted to keep the two interface implementations separate.